File: test/features/achievements/achievements_repository_test.dart
import 'dart:async';
import 'dart:collection';

import 'package:clashkingapp/core/services/api_service.dart';
import 'package:clashkingapp/core/config/api_config.dart';
import 'package:clashkingapp/features/achievements/data/achievements_repository.dart';
import 'package:clashkingapp/features/achievements/models/achievement.dart';
import 'package:clashkingapp/features/auth/data/auth_service.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:shared_preferences/shared_preferences.dart';

void main() {
  TestWidgetsFlutterBinding.ensureInitialized();

  setUp(() => SharedPreferences.setMockInitialValues({}));

  test('does not expose catalog entries before an authenticated response', () {
    final repository = AchievementsRepository(apiService: _FakeApiService());

    expect(repository.achievements, isEmpty);
  });

  test(
    'check uses the authenticated endpoint and maps all achievement states',
    () async {
      final api = _FakeApiService();
      final repository = AchievementsRepository(apiService: api);

      await repository.check();

      expect(api.lastEndpoint, '/achievements/check');
      expect(api.lastRequiresAuth, isTrue);
      expect(repository.achievements, hasLength(4));
      expect(
        repository.achievements
            .firstWhere((item) => item.id == AchievementId.townhall18)
            .earnedCount,
        3,
      );
      expect(
        repository.achievements
            .firstWhere((item) => item.id == AchievementId.warWarrior)
            .isUnlocked,
        isTrue,
      );
      expect(
        repository.achievements
            .firstWhere((item) => item.id == AchievementId.mrLegend)
            .isUnlocked,
        isFalse,
      );
      expect(
        repository.achievements
            .firstWhere((item) => item.id == AchievementId.defenseDoesntMatter)
            .modelUrl,
        endsWith('bad-legends-achievement-badge.glb'),
      );
    },
  );

  test(
    'load uses the catalog endpoint without running an award check',
    () async {
      final api = _FakeApiService();
      final repository = AchievementsRepository(apiService: api);

      await repository.load();

      expect(api.lastEndpoint, '/achievements');
      expect(api.lastRequiresAuth, isTrue);
    },
  );

  test('clear removes account-scoped state and notifies listeners', () async {
    final repository = AchievementsRepository(apiService: _FakeApiService());
    await repository.check();
    var notifications = 0;
    repository.addListener(() => notifications++);

    repository.clear();

    expect(repository.achievements, isEmpty);
    expect(repository.isRefreshing, isFalse);
    expect(notifications, 1);
  });

  test('binding auth clears achievements when the session changes', () async {
    final api = _FakeApiService();
    final repository = AchievementsRepository(apiService: api);
    await repository.check();
    final auth = AuthService(
      apiService: api,
      environment: ApiEnvironment.local,
    );
    repository.bindAuth(auth);
    repository.bindAuth(auth);

    await auth.initializeAuth();

    expect(repository.achievements, isEmpty);
    repository.dispose();
    auth.dispose();
  });

  test('a stale request cannot clear a newer refresh state', () async {
    final api = _FakeApiService();
    final first = Completer<Map<String, dynamic>>();
    final second = Completer<Map<String, dynamic>>();
    api.postResponses.addAll([first, second]);
    final repository = AchievementsRepository(apiService: api);

    final firstCheck = repository.check();
    repository.clear();
    final secondCheck = repository.check();
    first.complete(_FakeApiService.response);
    await firstCheck;

    expect(repository.isRefreshing, isTrue);
    second.complete(_FakeApiService.response);
    await secondCheck;
    expect(repository.isRefreshing, isFalse);
  });
}

class _FakeApiService extends ApiService {
  final Queue<Completer<Map<String, dynamic>>> postResponses = Queue();
  String? lastEndpoint;
  bool? lastRequiresAuth;

  static const response = <String, dynamic>{
    'items': <Map<String, dynamic>>[
      {
        'id': 'townhall_18',
        'asset_url':
            'https://assets.clashk.ing/achievements/town-hall-18-achievement-badge.glb',
        'repeatable': true,
        'earned_count': 3,
      },
      {
        'id': 'war_warrior',
        'asset_url':
            'https://assets.clashk.ing/achievements/war-champion-achievement-badge.glb',
        'repeatable': true,
        'earned_count': 1,
      },
      {
        'id': 'mr_legend',
        'asset_url':
            'https://assets.clashk.ing/achievements/perfect-legends-day-achievement-badge.glb',
        'repeatable': true,
        'earned_count': 0,
      },
      {
        'id': 'defense_doesnt_matter',
        'asset_url':
            'https://assets.clashk.ing/achievements/bad-legends-achievement-badge.glb',
        'repeatable': true,
        'earned_count': 0,
      },
    ],
  };

  @override
  Future<Map<String, dynamic>> get(
    String endpoint, {
    bool requiresAuth = true,
  }) async {
    lastEndpoint = endpoint;
    lastRequiresAuth = requiresAuth;
    if (endpoint == '/auth/me') {
      return const {
        'user_id': 'achievement-test-user',
        'discord_username': 'Achievement Tester',
        'avatar_url': '',
        'auth_methods': <String>['discord'],
      };
    }
    return response;
  }

  @override
  Future<Map<String, dynamic>> post(
    String endpoint,
    Map<String, String> body, {
    bool requiresAuth = false,
  }) async {
    lastEndpoint = endpoint;
    lastRequiresAuth = requiresAuth;
    if (postResponses.isNotEmpty) return postResponses.removeFirst().future;
    return response;
  }
}

File: expo/src/features/achievements/data/achievements-repository.test.ts
import { AchievementsCheckEndpoint } from '@clashking/api-contracts/expo';
import { Effect } from 'effect';

import type { ContractApiService } from '../../../core/api/contract-api';
import { AchievementsRepository } from './achievements-repository';

const response = {
  items: [
    {
      id: 'townhall_18',
      asset_url: 'https://assets.clashk.ing/achievements/town-hall-18-achievement-badge.glb',
      repeatable: true,
      earned_count: 3,
    },
    {
      id: 'war_warrior',
      asset_url: 'https://assets.clashk.ing/achievements/war-champion-achievement-badge.glb',
      repeatable: true,
      earned_count: 1,
    },
    {
      id: 'mr_legend',
      asset_url: 'https://assets.clashk.ing/achievements/perfect-legends-day-achievement-badge.glb',
      repeatable: true,
      earned_count: 0,
    },
    {
      id: 'defense_doesnt_matter',
      asset_url: 'https://assets.clashk.ing/achievements/bad-legends-achievement-badge.glb',
      repeatable: true,
      earned_count: -4,
    },
  ],
};

function setup(result: Record<string, unknown> | Promise<Record<string, unknown>> = response) {
  const execute = jest.fn(() => Effect.promise(() => Promise.resolve(result))) as unknown as
    jest.MockedFunction<ContractApiService['execute']>;
  const executeStatus = jest.fn() as unknown as ContractApiService['executeStatus'];
  return {
    execute,
    repository: new AchievementsRepository({ execute, executeStatus }),
  };
}

test('does not expose fallback entries before an authenticated response', () => {
  expect(setup().repository.snapshot.achievements).toEqual([]);
});

test('check posts an empty body with auth and maps server state in catalog order', async () => {
  const { repository, execute } = setup();
  await repository.check();
  expect(execute).toHaveBeenCalledWith(AchievementsCheckEndpoint, {
    path: {}, query: {}, body: {},
  });
  expect(repository.snapshot.achievements.map((item) => [item.id, item.earnedCount])).toEqual([
    ['townhall_18', 3],
    ['war_warrior', 1],
    ['mr_legend', 0],
    ['defense_doesnt_matter', 0],
  ]);
  expect(repository.snapshot.isRefreshing).toBe(false);
});

test('load uses the only deployed catalog response', async () => {
  const { repository, execute } = setup();
  await repository.load();
  expect(execute).toHaveBeenCalledWith(AchievementsCheckEndpoint, {
    path: {}, query: {}, body: {},
  });
});

test('skips invalid and unknown entries and rejects a missing items list', async () => {
  const { repository } = setup({
    items: [
      response.items[0],
      { id: 'unknown', asset_url: 'x', earned_count: 1, repeatable: true },
    ],
  });
  await repository.check();
  expect(repository.snapshot.achievements.map((item) => item.id)).toEqual(['townhall_18']);

  const malformed = setup({}).repository;
  await expect(malformed.check()).rejects.toBeInstanceOf(TypeError);
  expect(malformed.snapshot.isRefreshing).toBe(false);
});

test('session changes clear state and stale checks cannot replace a newer refresh', async () => {
  let resolveFirst: ((value: Record<string, unknown>) => void) | undefined;
  let resolveSecond: ((value: Record<string, unknown>) => void) | undefined;
  const first = new Promise<Record<string, unknown>>((resolve) => (resolveFirst = resolve));
  const second = new Promise<Record<string, unknown>>((resolve) => (resolveSecond = resolve));
  const execute = jest
    .fn()
    .mockImplementationOnce(() => Effect.promise(() => first))
    .mockImplementationOnce(() => Effect.promise(() => second)) as unknown as
    jest.MockedFunction<ContractApiService['execute']>;
  const executeStatus = jest.fn() as unknown as ContractApiService['executeStatus'];
  const repository = new AchievementsRepository({ execute, executeStatus });

  const stale = repository.check();
  repository.bindSession('user-2');
  const current = repository.check();
  resolveFirst?.(response);
  await stale;
  expect(repository.snapshot.isRefreshing).toBe(true);
  expect(repository.snapshot.achievements).toEqual([]);
  resolveSecond?.(response);
  await current;
  expect(repository.snapshot.isRefreshing).toBe(false);
  expect(repository.snapshot.achievements).toHaveLength(4);
});

test('coalesces checks while a refresh is active', async () => {
  let resolve: ((value: Record<string, unknown>) => void) | undefined;
  const pending = new Promise<Record<string, unknown>>((done) => (resolve = done));
  const { repository, execute } = setup(pending);
  const first = repository.check();
  await repository.check();
  expect(execute).toHaveBeenCalledTimes(1);
  resolve?.(response);
  await first;
});
